### ISSUE TITLE
Missing AOT'ed assemblies - Regression in VS 2022 17.1.0-pre2

### ISSUE THREAD

**Issue by tranb3r** [xamarin/xamarin-android#6615]
### Android application type

Classic Xamarin.Android (MonoAndroid12.0, etc.)

### Affected platform version

VS 2022 17.1.0-pre2

### Description

Using full aot, AOT'ed assemblies are missing in the apk.
This was working in previous previews of VS 2022 17.1.0.

### Steps to Reproduce

Enable full AOT and build an apk.
Inspect lib folder inside the built apk.

### Did you find any workaround?

No

**Comment by dellis1972** [comment]
@tranb3r  do you have a full diagnostic build log for this issue? 
I just tried to build a project with Full AOT against on latest `main` branch and I get the AOT'd files in the apk. 

It might also be useful to provide the settings for the configuration you are building for so we can try to replicate your exact issue.

**Comment by tranb3r** [comment]
Here are the logs.
[build.zip](https://github.com/xamarin/xamarin-android/files/7823340/build.zip)

build-ko.log is the one were aot'ed assemblies are missing.
build-ok.log is ok and was built on the same machine with VS 2022 17.0.4.

Thanks for investigating, and please ask if you need more information.

**Comment by dellis1972** [comment]
@tranb3r ok this is a problem

```
obj\Release\120\aot\arm64-v8a\libaot-System.Text.Encoding.CodePages.dll.so is an empty (no executable code) AOT assembly, not including it in the archive 
```

@grendello  has there been any bugs recently in the code which skips aot'd .so files when adding them to the apk?

**Comment by dellis1972** [comment]
@tranb3r are you able to provide us a zip containing the AOT'd  .so files? There is something about them which is tripping up our code which checks to see if they are empty.

**Comment by tranb3r** [comment]
Here is a zip with the aot'ed so files:
[arm64-v8a.zip](https://github.com/xamarin/xamarin-android/files/7824311/arm64-v8a.zip)

**Comment by tranb3r** [comment]
I've tested the ELFHelper.IsEmptyAOTLibrary method on my so files, the result is that all libs are considered as empty (which is obviously incorrect).

**Comment by tranb3r** [comment]
I'm actually using these arguments:
`<AndroidAotAdditionalArguments>no-write-symbols,nodebug</AndroidAotAdditionalArguments>`
When I remove this from my csproj, there is no more issue. The no-write-symbols arg probably messes with your IsEmptyAOTLibrary code.
The workaround is probably to remove the no-write-symbols arg, but I'd rather a proper fix, since I do not think this new functionality of checking aot libs for emptyness is really useful when doing full aot.

**Comment by dellis1972** [comment]
@tranb3r thanks for the detailed info. We will to get our native expert @grendello  to take a look at this next week. Glad you found a workaround in the mean time.

**Comment by jonathanpeppers** [comment]
We might consider passing `no-write-symbols,nodebug` by default. Wouldn't that improve .apk size for AOT?

**Comment by tranb3r** [comment]
> We might consider passing `no-write-symbols,nodebug` by default. Wouldn't that improve .apk size for AOT?

Yes, the impact of these arguments on apk size is huge (especially when doing full aot).
It's why the workaround is problematic : at least 10% increase of apk size on my app when removing no-write-symbols arg.

**Comment by grendello** [comment]
> @tranb3r ok this is a problem
> 
> ```
> obj\Release\120\aot\arm64-v8a\libaot-System.Text.Encoding.CodePages.dll.so is an empty (no executable code) AOT assembly, not including it in the archive 
> ```
> 
> @grendello has there been any bugs recently in the code which skips aot'd .so files when adding them to the apk?

nope, I haven't seen any. The code added in db161ae7eaad1acb344192614eab9ebf14896291 should be skipping only DSOs that are empty, but the decision on that is based on certain assumptions, so bugs are possible.

**Comment by grendello** [comment]
> I'm actually using these arguments: `<AndroidAotAdditionalArguments>no-write-symbols,nodebug</AndroidAotAdditionalArguments>` When I remove this from my csproj, there is no more issue. The no-write-symbols arg probably messes with your IsEmptyAOTLibrary code. The workaround is probably to remove the no-write-symbols arg, but I'd rather a proper fix, since I do not think this new functionality of checking aot libs for emptyness is really useful when doing full aot.

This is the real bug. `IsEmptyAOTLibrary` should let the assembly through if there are no symbols. I'll look into it this week.

**Comment by tranb3r** [comment]
> > I'm actually using these arguments: `<AndroidAotAdditionalArguments>no-write-symbols,nodebug</AndroidAotAdditionalArguments>` When I remove this from my csproj, there is no more issue. The no-write-symbols arg probably messes with your IsEmptyAOTLibrary code. The workaround is probably to remove the no-write-symbols arg, but I'd rather a proper fix, since I do not think this new functionality of checking aot libs for emptyness is really useful when doing full aot.
> 
> This is the real bug. `IsEmptyAOTLibrary` should let the assembly through if there are no symbols. I'll look into it this week.

There are symbols. But not of Function type. Most of them are actually of NotSpecified type or Section type.

**Comment by grendello** [comment]
> > > I'm actually using these arguments: `<AndroidAotAdditionalArguments>no-write-symbols,nodebug</AndroidAotAdditionalArguments>` When I remove this from my csproj, there is no more issue. The no-write-symbols arg probably messes with your IsEmptyAOTLibrary code. The workaround is probably to remove the no-write-symbols arg, but I'd rather a proper fix, since I do not think this new functionality of checking aot libs for emptyness is really useful when doing full aot.
> > 
> > 
> > This is the real bug. `IsEmptyAOTLibrary` should let the assembly through if there are no symbols. I'll look into it this week.
> 
> There are symbols. But not of Function type. Most of them are actually of NotSpecified type or Section type.

Which makes the code consider the DSO as empty, that is containing no code. From a quick look, it seems I need to make the code check for presence of any non-zero size symbol with the `_plt` suffix (which is a jump table to non-public symbols in the DSO)

**Comment by tranb3r** [comment]
I've just tested VS 2022 17.1.0-pre3. I'm still having this issue.

**Comment by grendello** [comment]
@tranb3r I don't know if the commit made it to that release, @jonpryor?

**Comment by tranb3r** [comment]
It would be so much easier if you published release notes...

**Comment by grendello** [comment]
@tranb3r Release notes can be found here https://docs.microsoft.com/en-us/xamarin/android/release-notes/12/12.2

**Comment by tranb3r** [comment]
> @tranb3r Release notes can be found here https://docs.microsoft.com/en-us/xamarin/android/release-notes/12/12.2

I know. But there is no release notes for xamarin.android 12.2.0.2 which is the version published with VS 2022 17.1.0-pre3

**Comment by tranb3r** [comment]
Come on guys... this is so frustrating !!

**Comment by jonpryor** [comment]
@tranb3r: We are attempting to get the fix in by Visual Studio 2022 17.1 GA, if not in an earlier preview.
